observer/baseNodeProvider.go

```diff
@@ -148,80 +207,119 @@ func (bnp *baseNodeProvider) isReceivedSyncedNodeExistent(receivedNode *data.Nod
 	return false
 }
 
+func (bnp *baseNodeProvider) isReceivedSyncedNodeExistentAsFallback(receivedNode *data.NodeData) bool {
+	for _, node := range bnp.syncedFallbackNodes {
+		if node.Address == receivedNode.Address && node.ShardId == receivedNode.ShardId {
+			return true
+		}
+	}
+
+	return false
+}
+
 func (bnp *baseNodeProvider) addToOutOfSyncUnprotected(node *data.NodeData) {
-	if bnp.outOfSyncNodes == nil {
-		bnp.outOfSyncNodes = make([]*data.NodeData, 0)
+	source := &bnp.outOfSyncNodes
+	if node.IsFallback {
+		source = &bnp.outOfSyncFallbackNodes
 	}
 
-	for _, oosNode := range bnp.outOfSyncNodes {
+	if *source == nil {
+		*source = make([]*data.NodeData, 0)
+	}
+
+	for _, oosNode := range *source {
 		if oosNode.Address == node.Address && oosNode.ShardId == node.ShardId {
 			return
 		}
 	}
 
-	bnp.outOfSyncNodes = append(bnp.outOfSyncNodes, node)
+	*source = append(*source, node)
 }
 
 func (bnp *baseNodeProvider) removeOutOfSyncNodesUnprotected(
 	outOfSyncNodes []*data.NodeData,
 	syncedNodesMap map[uint32][]*data.NodeData,
+	syncedFallbackNodesMap map[uint32][]*data.NodeData,
 ) {
 	if len(outOfSyncNodes) == 0 {
 		bnp.outOfSyncNodes = make([]*data.NodeData, 0)
+		bnp.outOfSyncFallbackNodes = make([]*data.NodeData, 0)
 		return
 	}
 
 	for _, outOfSyncNode := range outOfSyncNodes {
-		if len(syncedNodesMap[outOfSyncNode.ShardId]) < 1 {
-			log.Warn("cannot remove observer as not enough will remain in shard",
+		hasOneSyncedNode := len(syncedNodesMap[outOfSyncNode.ShardId]) >= 1
+		hasEnoughSyncedFallbackNodes := len(syncedFallbackNodesMap[outOfSyncNode.ShardId]) > 1
+		canDeleteFallbackNode := hasOneSyncedNode || hasEnoughSyncedFallbackNodes
+		if outOfSyncNode.IsFallback && canDeleteFallbackNode {
+			bnp.removeNodeUnprotected(outOfSyncNode)
+			continue
+		}
+
+		// if trying to delete last fallback, use last known synced node
+		// if backup node does not exist, keep fallback
+		hasBackup := bnp.lastSyncedNodes[outOfSyncNode.ShardId] != nil
+		if outOfSyncNode.IsFallback && hasBackup {
+			bnp.removeNodeUnprotected(outOfSyncNode)
+			continue
+		}
+
+		hasEnoughSyncedNodes := len(syncedNodesMap[outOfSyncNode.ShardId]) >= 1
+		if hasEnoughSyncedNodes {
+			bnp.removeNodeUnprotected(outOfSyncNode)
+			continue
+		}
+
+		// trying to remove last synced node
+		// if fallbacks are available, save this one as backup and use fallbacks
+		// else, keep using this one
+		// save this last regular observer as backup in case fallbacks go offline
+		// also, if this is the old fallback observer which didn't get synced, keep it in list
+		wasSyncedAtPreviousStep := bnp.isReceivedSyncedNodeExistent(outOfSyncNode)
+		isBackupObserver := bnp.lastSyncedNodes[outOfSyncNode.ShardId] == outOfSyncNode
+		if !outOfSyncNode.IsFallback && wasSyncedAtPreviousStep || isBackupObserver {
```

Comment:
can you extract `!outOfSyncNode.IsFallback && wasSyncedAtPreviousStep` into a new variable?